File: android/src/main/java/com/anuvaya/plex/PlexModule.kt
package com.anuvaya.plex

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import android.content.pm.PackageManager
import com.google.android.play.core.appupdate.AppUpdateManager
import com.google.android.play.core.appupdate.AppUpdateManagerFactory
import com.google.android.play.core.install.model.AppUpdateType
import com.google.android.play.core.install.model.UpdateAvailability
import com.google.android.play.core.install.InstallStateUpdatedListener
import com.google.android.play.core.install.model.InstallStatus

class PlexModule : Module() {
  private val paymentApps = listOf(
    PaymentAppInfo("com.google.android.apps.nbu.paisa.user", "Google Pay", "tez"),
    PaymentAppInfo("com.phonepe.app", "PhonePe", "phonepe"),
    PaymentAppInfo("net.one97.paytm", "Paytm", "paytmmp"),
    PaymentAppInfo("com.dreamplug.androidapp", "CRED", "credpay"),
    PaymentAppInfo("com.mobikwik_new", "MobiKwik", "mobikwik"),
    PaymentAppInfo("com.freecharge.android", "FreeCharge", "freecharge"),
    PaymentAppInfo("in.fampay.app", "FamPay", "in.fampay.app"),
    PaymentAppInfo("in.org.npci.upiapp", "BHIM", "bhim"),
    PaymentAppInfo("in.amazon.mShop.android.shopping", "Amazon Pay", "amazonpay"),
    PaymentAppInfo("com.naviapp", "Navi", "navi"),
    PaymentAppInfo("com.axis.mobile", "Axis Mobile", "kiwi"),
    PaymentAppInfo("com.enstage.wibmo.hdfc", "PayZapp", "payzapp"),
    PaymentAppInfo("money.jupiter.app", "Jupiter", "jupiter"),
    PaymentAppInfo("com.csam.icici.bank.imobile", "iMobile Pay", "icici"),
    PaymentAppInfo("com.sbi.lotza.mbanking", "YONO SBI", "sbiyono"),
    PaymentAppInfo("com.jio.myjio", "MyJio", "myjio"),
    PaymentAppInfo("com.sliceit.android", "Slice", "slice-upi"),
    PaymentAppInfo("com.barodampay.app", "Bank of Baroda UPI", "bobupi"),
    PaymentAppInfo("com.whatsapp", "WhatsApp", "whatsapp")
  )

  data class PaymentAppInfo(
    val packageName: String,
    val displayName: String,
    val scheme: String
  )

  override fun definition() = ModuleDefinition {
    Name("Plex")

    AsyncFunction("getPaymentContext") {
      val installedApps = detectInstalledPaymentApps()
      mapOf(
        "installedApps" to installedApps,
        "totalAppsChecked" to paymentApps.size,
        "detectionMethod" to "package_manager"
      )
    }

    AsyncFunction("checkForUpdate") { options: Map<String, Any?>?, promise: Promise ->
      val context = appContext.reactContext
      if (context == null) {
        promise.resolve(
          mapOf(
            "platform" to "android",
            "isAvailable" to false,
            "recommendedType" to "none",
            "localVersion" to ""
          )
        )
        return@AsyncFunction
      }

      val pm = context.packageManager
      val localVersion = try {
        val pInfo = pm.getPackageInfo(context.packageName, 0)
        pInfo.versionName ?: ""
      } catch (e: Exception) { "" }

      val manager: AppUpdateManager = AppUpdateManagerFactory.create(context)
      val task = manager.appUpdateInfo
      task.addOnSuccessListener { info ->
        val availability = info.updateAvailability()
        var recommended = "none"
        var isAvailable = false
        if (availability == UpdateAvailability.UPDATE_AVAILABLE) {
          isAvailable = true
          recommended = when {
            info.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE) -> "immediate"
            info.isUpdateTypeAllowed(AppUpdateType.FLEXIBLE) -> "flexible"
            else -> "none"
          }
        }

        val data = mutableMapOf<String, Any?>(
          "platform" to "android",
          "isAvailable" to isAvailable,
          "recommendedType" to recommended,
          "localVersion" to localVersion,
        )

        // Provide version code if available
        try {
          val code = info.availableVersionCode()
          data["remoteVersion"] = code
        } catch (_: Exception) {}

        promise.resolve(data)
      }
      task.addOnFailureListener { e ->
        promise.resolve(
          mapOf(
            "platform" to "android",
            "isAvailable" to false,
            "recommendedType" to "none",
            "localVersion" to localVersion
          )
        )
      }
    }

    AsyncFunction("startUpdate") { options: Map<String, Any?>?, promise: Promise ->
      val context = appContext.currentActivity ?: appContext.reactContext
      val activity = appContext.currentActivity
      if (context == null || activity == null) {
        promise.resolve(mapOf("started" to false))
        return@AsyncFunction
      }

      val manager: AppUpdateManager = AppUpdateManagerFactory.create(context)
      val task = manager.appUpdateInfo
      task.addOnSuccessListener { info ->
        if (info.updateAvailability() != UpdateAvailability.UPDATE_AVAILABLE) {
          promise.resolve(mapOf("started" to false))
          return@addOnSuccessListener
        }

        val requestedType = when ((options?.get("type") as? String)?.lowercase()) {
          "immediate" -> AppUpdateType.IMMEDIATE
          "flexible" -> AppUpdateType.FLEXIBLE
          else -> null
        }

        val type = when {
          requestedType != null && info.isUpdateTypeAllowed(requestedType) -> requestedType
          info.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE) -> AppUpdateType.IMMEDIATE
          info.isUpdateTypeAllowed(AppUpdateType.FLEXIBLE) -> AppUpdateType.FLEXIBLE
          else -> null
        }

        if (type == null) {
          promise.resolve(mapOf("started" to false))
          return@addOnSuccessListener
        }

        if (type == AppUpdateType.FLEXIBLE) {
          // For flexible updates, auto-complete when downloaded
          // Use object expression so we can unregister self
          val objListener = object : InstallStateUpdatedListener {
            override fun onStateUpdate(state: com.google.android.play.core.install.InstallState) {
              when (state.installStatus()) {
                InstallStatus.DOWNLOADED -> {
                  manager.completeUpdate()
                  manager.unregisterListener(this)
                }
                InstallStatus.INSTALLED, InstallStatus.FAILED, InstallStatus.CANCELED -> {
                  manager.unregisterListener(this)
                }
                else -> {}
              }
            }
          }
          manager.registerListener(objListener)
        }

        try {
          // Request code can be any integer; we don't need the result here
          val REQUEST_CODE = 17362
          manager.startUpdateFlowForResult(
            info,
            type,
            activity,
            REQUEST_CODE
          )
          promise.resolve(mapOf("started" to true))
        } catch (e: Exception) {
          promise.resolve(mapOf("started" to false))
        }
      }
      task.addOnFailureListener {
        promise.resolve(mapOf("started" to false))
      }
    }

  }

  private fun detectInstalledPaymentApps(): List<Map<String, Any>> {
    val installedApps = mutableListOf<Map<String, Any>>()
    val packageManager = appContext.reactContext?.packageManager ?: return installedApps

    for (app in paymentApps) {
      val isInstalled = try {
        packageManager.getPackageInfo(app.packageName, PackageManager.GET_META_DATA)
        true
      } catch (e: PackageManager.NameNotFoundException) {
        false
      }

      val appInfo = mapOf(
        "scheme" to app.scheme,
        "name" to app.displayName,
        "isInstalled" to isInstalled,
        "packageName" to app.packageName
      )

      installedApps.add(appInfo)
    }

    return installedApps
  }
}

File: ios/PlexModule.swift
import ExpoModulesCore
import UIKit

public class PlexModule: Module {
  private let paymentApps: [(scheme: String, name: String)] = [
    ("tez", "Google Pay"),
    ("phonepe", "PhonePe"),
    ("paytmmp", "Paytm"),
    ("credpay", "CRED"),
    ("mobikwik", "MobiKwik"),
    ("freecharge", "FreeCharge"),
    ("in.fampay.app", "FamPay"),
    ("bhim", "BHIM"),
    ("amazonpay", "Amazon Pay"),
    ("navi", "Navi"),
    ("kiwi", "Kiwi"),
    ("payzapp", "PayZapp"),
    ("jupiter", "Jupiter"),
    ("omnicard", "Omni Card"),
    ("icici", "iMobile Pay"),
    ("popclubapp", "PopClub"),
    ("sbiyono", "YONO SBI"),
    ("myjio", "MyJio"),
    ("slice-upi", "Slice"),
    ("bobupi", "Bank of Baroda UPI"),
    ("shriramone", "Shriram One"),
    ("indusmobile", "IndusInd Bank"),
    ("whatsapp", "WhatsApp"),
  ]

  public func definition() -> ModuleDefinition {
    Name("Plex")

    AsyncFunction("getPaymentContext") { () -> [String: Any] in
      let installedApps = self.detectInstalledPaymentApps()
      return [
        "installedApps": installedApps,
        "totalAppsChecked": self.paymentApps.count,
        "detectionMethod": "url_scheme",
      ]
    }

    AsyncFunction("checkForUpdate") { (options: [String: Any]?, promise: Promise) in
      let localVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
      let optAppId = options?["appStoreId"] as? String
      let optBundle = options?["bundleId"] as? String
      let country = options?["countryCode"] as? String

      var components = URLComponents(string: "https://itunes.apple.com/lookup")
      var queryItems: [URLQueryItem] = []
      if let appId = optAppId, !appId.isEmpty {
        queryItems.append(URLQueryItem(name: "id", value: appId))
      } else if let bundleOverride = optBundle, !bundleOverride.isEmpty {
        queryItems.append(URLQueryItem(name: "bundleId", value: bundleOverride))
      } else if let bundleId = Bundle.main.bundleIdentifier {
        queryItems.append(URLQueryItem(name: "bundleId", value: bundleId))
      }
      if let country = country, !country.isEmpty {
        queryItems.append(URLQueryItem(name: "country", value: country))
      }
      components?.queryItems = queryItems
      guard let url = components?.url else {
        promise.resolve([
          "platform": "ios",
          "isAvailable": false,
          "recommendedType": "none",
          "localVersion": localVersion,
        ])
        return
      }

      URLSession.shared.dataTask(with: url) { data, _, error in
        guard error == nil, let data = data else {
          promise.resolve([
            "platform": "ios",
            "isAvailable": false,
            "recommendedType": "none",
            "localVersion": localVersion,
          ])
          return
        }

        var remoteVersion: String? = nil
        var storeUrl: String? = nil
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let results = json["results"] as? [[String: Any]],
           let first = results.first {
          if let v = first["version"] as? String { remoteVersion = v }
          if let u = first["trackViewUrl"] as? String { storeUrl = u }
        }

        let isAvailable: Bool = {
          guard let remote = remoteVersion else { return false }
          return self.isRemoteVersionNewer(remote, than: localVersion)
        }()

        var result: [String: Any] = [
          "platform": "ios",
          "isAvailable": isAvailable,
          "recommendedType": isAvailable ? "store" : "none",
          "localVersion": localVersion,
        ]
        if let remoteVersion = remoteVersion { result["remoteVersion"] = remoteVersion }
        if let storeUrl = storeUrl { result["storeUrl"] = storeUrl }
        promise.resolve(result)
      }.resume()
    }

    AsyncFunction("startUpdate") { (options: [String: Any]?, promise: Promise) in
      let optAppId = options?["appStoreId"] as? String
      let optBundle = options?["bundleId"] as? String
      let country = (options?["countryCode"] as? String)?.lowercased()

      // If appStoreId is provided, construct an itms-apps URL directly (more reliable)
      if let appId = optAppId, !appId.isEmpty {
        let countryPath = (country != nil && !country!.isEmpty) ? "/\(country!)" : ""
        let itmsUrl = URL(string: "itms-apps://apps.apple.com\(countryPath)/app/id\(appId)")
        let httpsUrl = URL(string: "https://apps.apple.com\(countryPath)/app/id\(appId)")
        DispatchQueue.main.async {
          if let u = itmsUrl {
            UIApplication.shared.open(u, options: [:]) { success in
              if success {
                promise.resolve(["started": true])
              } else if let alt = httpsUrl {
                UIApplication.shared.open(alt, options: [:]) { success2 in
                  promise.resolve(["started": success2])
                }
              } else {
                promise.resolve(["started": false])
              }
            }
          } else if let alt = httpsUrl {
            UIApplication.shared.open(alt, options: [:]) { success in
              promise.resolve(["started": success])
            }
          } else {
            promise.resolve(["started": false])
          }
        }
        return
      }

      // Otherwise, lookup by bundleId and open the trackViewUrl
      var components = URLComponents(string: "https://itunes.apple.com/lookup")
      var queryItems: [URLQueryItem] = []
      if let bundleOverride = optBundle, !bundleOverride.isEmpty {
        queryItems.append(URLQueryItem(name: "bundleId", value: bundleOverride))
      } else if let bundleId = Bundle.main.bundleIdentifier {
        queryItems.append(URLQueryItem(name: "bundleId", value: bundleId))
      }
      if let country = country, !country.isEmpty {
        queryItems.append(URLQueryItem(name: "country", value: country))
      }
      components?.queryItems = queryItems
      guard let url = components?.url else {
        promise.resolve(["started": false])
        return
      }

      URLSession.shared.dataTask(with: url) { data, _, error in
        guard error == nil, let data = data,
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let results = json["results"] as? [[String: Any]],
              let first = results.first,
              let storeUrlStr = first["trackViewUrl"] as? String else {
          promise.resolve(["started": false])
          return
        }

        // Try itms-apps first, fallback to https
        let transformed = storeUrlStr.replacingOccurrences(of: "https://", with: "itms-apps://")
        let itmsUrl = URL(string: transformed)
        let httpsUrl = URL(string: storeUrlStr)

        DispatchQueue.main.async {
          if let u = itmsUrl {
            UIApplication.shared.open(u, options: [:]) { success in
              if success {
                promise.resolve(["started": true])
              } else if let alt = httpsUrl {
                UIApplication.shared.open(alt, options: [:]) { success2 in
                  promise.resolve(["started": success2])
                }
              } else {
                promise.resolve(["started": false])
              }
            }
          } else if let alt = httpsUrl {
            UIApplication.shared.open(alt, options: [:]) { success in
              promise.resolve(["started": success])
            }
          } else {
            promise.resolve(["started": false])
          }
        }
      }.resume()
    }
  }

  private func detectInstalledPaymentApps() -> [[String: Any]] {
    var installedApps: [[String: Any]] = []

    for app in paymentApps {
      let scheme = app.scheme
      let name = app.name

      // Create URL with scheme
      guard let url = URL(string: "\(scheme)://") else {
        continue
      }

      // Check if app can handle this URL scheme
      let isInstalled = UIApplication.shared.canOpenURL(url)

      let appInfo: [String: Any] = [
        "scheme": scheme,
        "name": name,
        "isInstalled": isInstalled,
        "packageName": scheme,  // Using scheme as identifier for now
      ]

      installedApps.append(appInfo)
    }

    return installedApps
  }

  // Compare dotted numeric version strings (e.g., "1.2.10" > "1.2.3")
  private func isRemoteVersionNewer(_ remote: String, than local: String) -> Bool {
    let r = remote.split(separator: ".").compactMap { Int($0) }
    let l = local.split(separator: ".").compactMap { Int($0) }
    let count = max(r.count, l.count)
    for i in 0..<count {
      let rv = i < r.count ? r[i] : 0
      let lv = i < l.count ? l[i] : 0
      if rv != lv { return rv > lv }
    }
    return false
  }
}
